### PR TITLE
Generating reactions programmatically using Reaction Systems

### PR DESCRIPTION
**Pull request by yewalenikhil65**
Tutorial for Implementation of (Smoluchowski coagulation equation) population balance equation- by programmatically generating the reactions using ReactionSystem
Based on discussion at https://github.com/SciML/Catalyst.jl/issues/292

### CODE REVIEW

**Comment by isaacsas** [comment]
@yewalenikhil65 I'll go through tomorrow morning and give you some comments. Got busy with meetings and class stuff today.

**Comment by yewalenikhil65** [comment]
> walenikhil65 I'll go through tomorrow morning and give you some comments. Got busy with meetings and class stuff today.

No problem ,Sir

**Comment by isaacsas** [comment]
You should be able to `git pull` to sync my changes back to your local git repository.

File: docs/src/tutorials/generating_reactions_programmatically.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+We begin by importing some necessary packages.
+```julia
+using ModelingToolkit, LinearAlgebra
+using DiffEqBase, DiffEqJump, OrdinaryDiffEq
+using LoopVectorization, Plots
+using BenchmarkTools
+using SpecialFunctions
+```
+Suppose the maximum cluster size is `N`. Lets initialize the system with some initial concentration `C`, initial number of singlets `uₒ` in the system. Since its a bimolecular chain of Reaction system(`nr` number of reactions), the bulk volume `V` of the system in which these binary collisions occur is important in the calculation of rate laws.
```
Comment (isaacsas) [inline docs/src/tutorials/generating_reactions_programmatically.md:14]: Should the initial concentration here be called `C` or `N_0`?

File: docs/src/tutorials/generating_reactions_programmatically.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  
+```julia
+## Parameter
+N = 10                       # maximum clusters size
+Vₒ = (4π/3)*(10e-06*100)^3   # volume of a singlet/monomer in cm³
```
Comment (isaacsas) [inline docs/src/tutorials/generating_reactions_programmatically.md:19]: Are singlets and monomers the same? If so, I suggest using monomer, dimer, trimer, etc throughout and not using signlet/doublet/triplet, etc.

File: docs/src/tutorials/generating_reactions_programmatically.md (hunk 1 above)
Comment (yewalenikhil65) [reply docs/src/tutorials/generating_reactions_programmatically.md:14]: You are right..made a mistake while typing.. Initial concentration should be `Nₒ`

File: docs/src/tutorials/generating_reactions_programmatically.md (hunk 2 above)
Comment (yewalenikhil65) [reply docs/src/tutorials/generating_reactions_programmatically.md:19]: > Are singlets and monomers the same?
>
Yes they are same..
 >If so, I suggest using monomer, dimer, trimer, etc throughout and not using signlet/doublet/triplet, etc.
>
I will make these changes

**Comment by yewalenikhil65** [comment]
Hey @isaacsas   updated according to your suggestions..

**Comment by isaacsas** [comment]
@yewalenikhil65 made some more updates. Slowly getting through it. (This week is pretty busy.)

I do have one question for you, see below.

**Comment by isaacsas** [comment]
Actually never mind; I forgot that I had figured out what I was confused about!

**Comment by yewalenikhil65** [comment]
Corrected a little mistake I spotted in writing `kv` for constant kernel case.. I like your changes in wording too

**Comment by isaacsas** [comment]
OK, I updated it one more time. I'd be happy to merge it now. Some notes:
- You weren't actually using ModelingToolkit / Catalyst in your previous code. When you created the `MassActionJump` and passed it to `JumpProblem` you were actually just using `DiffEqJump` directly. 
- Along those lines, the `ReactionSystem` model used `2*k[n]` for reactions with involving reactants that had the same number of monomers in them, however, what you were plotting and matching to the theory was just using `k[n]` for such reactions! I switched the `ReactionSystem` to use `k[n]` too since that is what made it match the theoretical formulas you used. Please double check that this is all correct though! You can see the new figure I made in the assets folder.

Let me know after you've taken a look, and if you approve I'll merge.

**Comment by yewalenikhil65** [comment]
> OK, I updated it one more time. I'd be happy to merge it now. Some notes:
> 
>     * You weren't actually using ModelingToolkit / Catalyst in your previous code. When you created the `MassActionJump` and passed it to `JumpProblem` you were actually just using `DiffEqJump` directly.
> 
Yes, you are right. But i guess pushing reactions into for loop uses ModelingToolkit .
>     * Along those lines, the `ReactionSystem` model used `2*k[n]` for reactions with involving reactants that had the same number of monomers in them, however, what you were plotting and matching to the theory was just using `k[n]` for such reactions! I switched the `ReactionSystem` to use `k[n]` too since that is what made it match the theoretical formulas you used. Please double check that this is all correct though! You can see the new figure I made in the assets folder.
> 
Again you are right.I plotted for just `k[n]` when I wrote the tutorial and uploaded that figure...But i have double checked using `2*k[n]` and `k[n]`also. In both cases, multiplicative kernel doesn't match with analytical ..I do not know why!
> 
> Let me know after you've taken a look, and if you approve I'll merge.
>
We can merge it now. Only bit I am unsure is how to reason no-match of multiplicative kernel case, so i skipped commenting on it.

**Comment by isaacsas** [comment]
Hmm, perhaps we should just use the additive kernel then and remove the other two?

**Comment by yewalenikhil65** [comment]
> Hmm, perhaps we should just use the additive kernel then and remove the other two?

Additive and constant kernels..Both are matching with analytical profiles.. 
I am not comfortable with removing multiplicative kernel even though its not matching. What do you think about calculation of the `bulk volume of the system` with which we are using in calculation of rate-constants ? Did we miss anything there ?

**Comment by yewalenikhil65** [comment]
Hey @isaacsas ..just realised something
This non-matching behaviour for `2*k[n]` that you got is because of this https://github.com/SciML/Catalyst.jl/pull/307/commits/58a703e4db8c7e23c750293281eced3212ebf69a
I believe if you do add `stoich construction` as I did previously ,it matches in additive and constant kernel even if you use `2*k[n]` as i mentioned earlier.I am confused why it is such though.

**Comment by isaacsas** [comment]
The `MassActionJump` you are building is using `k[n]` though not `2*k[n]` for those reactions, right? When you use the `mass_act_jump` none of the `ReactionSystem` code is even getting used so it is irrelevant what you set as the rate in the `Reaction`s.

**Comment by isaacsas** [comment]
Unfortunately, I haven't read this exact solution paper / method, so can't comment on why the multiplicative case does not match it.

**Comment by yewalenikhil65** [comment]
> The `MassActionJump` you are building is using `k[n]` though not `2*k[n]` for those reactions, right? When you use the `mass_act_jump` none of the `ReactionSystem` code is even getting used so it is irrelevant what you set as the rate in the `Reaction`s.

Ahh.. understood now.. In that case, https://github.com/SciML/Catalyst.jl/commit/58a703e4db8c7e23c750293281eced3212ebf69a seems correct and usage of `k[n]` seems more correct. I just `latexified` reactions into ODEs and checked if it matches with equation in wiki page.Usage of `k[n]` is appropriate one as you have done. 
I think we can leave multiplicative case for now then. and merge only additive and constant kernels case ?

**Comment by isaacsas** [comment]
OK, can you remove the multiplicative case from the tutorial text and code? I'll then take a look and merge.

**Comment by yewalenikhil65** [comment]
> OK, can you remove the multiplicative case from the tutorial text and code? I'll then take a look and merge.

@isaacsas ..Done!

**Review by isaacsas — APPROVED** [review]

**Comment by isaacsas** [comment]
All merged! Thanks!

**Pull request merged**